### ISSUE TITLE
[JENKINS-33151] GitHub commit status design is bad

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/github-branch-source-plugin#932]
<p>The latest release of the GitHub Branch Source Plugin added support for github commit status notifications. I was seriously happy about this, because I thought this would allow me to remove generic </p>

<pre>
step([$class: <span class="code-quote">'GitHubSetCommitStatusBuilder'</span>])
</pre>
<p>and</p>

<pre>
step([$class: <span class="code-quote">'GitHubCommitNotifier'</span>, resultOnFailure: <span class="code-quote">'FAILURE'</span>])
</pre>
<p>steps from my workflow scripts.</p>

<p>Unfortunately, I was seriously let down by the new functionality, and I found major issues with how it interacts with the workflow plugin.</p>

<h4><a name="Nosupportformultiple%60node%60%2F%60checkout%60combinations"></a><b>No support for multiple `node` / `checkout` combinations</b></h4>

<p>  My Workflow scripts reserve different nodes and checkout the same code on<br/>
  these. This might happen in parallel, or in sequence. After the steps on the<br/>
  first node finish, the build gets marked as successful, even if later steps<br/>
  are still pending. If a later step fails, a previously successful build will<br/>
  suddenly get marked as successful. I haven't tested with parallel builds yet,<br/>
  but I believe that might cause even weirder issues.</p>

<h4><a name="Noflexibility"></a>No flexibility</h4>

<p>  My workflow jobs sometimes consist of multiple different steps,<br/>
  like Validation, Build, Test, Deployments. With the generic steps mentioned<br/>
  above, I was able to be flexible on when the "final" commit status was set.<br/>
  E.g for one job, I want the commit status to be set after the code was build<br/>
  and tested, irrelevant of the deployment step.</p>

<p>I thus propose an alternative solution:</p>

<p>Instead of making the sending of the commit status automatic, custom workflow<br/>
steps should be added to the `github` plugin. These steps should be flexible<br/>
enough to allow users to use them in even the most complex workflow scripts,<br/>
while also having sensible defaults to make them easy to use.</p>

<h4><a name="%7B%7BwithGitHubCommitStatus%7D%7D"></a><tt>withGitHubCommitStatus</tt></h4>

<p>  This step should take a block and update the commit status on GitHub. By default,<br/>
  it should use the GitHub URL, Commit ID and other required information from<br/>
  the main git source for the job.</p>

<p>  This step could take a bunch of optional parameters:</p>

<ul>
	<li><tt>sha</tt>: Commit ID to set the status for. (Default: Taken from build)</li>
	<li><tt>repo</tt>: Repository name. (Default: Taken from build)</li>
	<li><tt>owner</tt>: Owner name (user or organization). (Default: Taken from build)</li>
	<li><tt>context</tt>: Context to be used for the status. (Default: "default" or "continuous-integration/jenkins")</li>
	<li><tt>targetUrl</tt>: URL to be linked to in the GitHub UI. (Default: Build URL)</li>
</ul>

<p>  Based on whether the contained block passes or not, the status on GitHub should be set<br/>
  accordingly.</p>

<h4><a name="%7B%7BsetGitHubCommitStatus%7D%7D"></a><tt>setGitHubCommitStatus</tt></h4>

<p>  This is basically a non-block version the same as <tt>withGitHubCommitStatus</tt>.</p>

<p>  It should take the same parameters as <tt>withGitHubCommitStatus</tt>, plus the following:</p>

<ul>
	<li><tt>state</tt>: <tt>pending</tt>, <tt>success</tt>, <tt>error</tt>, or <tt>failure</tt>.</li>
	<li><tt>description</tt>: A custom description.</li>
</ul>

<p>Here's same example Workflow scripts that show how this could be used:</p>

<pre>
#!/usr/bin/env groovy

stage name: <span class="code-quote">'build'</span>

parallel lint: {
  withGitHubCommitStatus context: <span class="code-quote">"continuous-integration/jenkins/lint"</span> {
    node {
      checkout scm

      sh(<span class="code-quote">'script/setup'</span>)
      sh(<span class="code-quote">'script/lint'</span>)
    }
  }
}, test: {
  withGitHubCommitStatus context: <span class="code-quote">"continuous-integration/jenkins/test"</span> {
    node {
      checkout scm

      sh(<span class="code-quote">'script/setup'</span>)
      sh(<span class="code-quote">'script/test'</span>)
    }
  }
}

<span class="code-comment">// Prepare a deployment
</span>stage name: <span class="code-quote">'prepare-deployment'</span>
node {
  checkout scm

  sh(<span class="code-quote">'script/setup'</span>)
  sh(<span class="code-quote">'script/prepare-deploy'</span>)
}

<span class="code-comment">// Run deployments
</span><span class="code-keyword">if</span> (currentBuild.result != <span class="code-quote">'FAILURE'</span> && env.BRANCH_NAME == <span class="code-quote">'master'</span>) {
  input <span class="code-quote">'Trigger Deployment?'</span>

  stage name: <span class="code-quote">'deployment'</span>, concurrency: 1

  node {
    checkout scm

    sh(<span class="code-quote">'script/setup'</span>)
    sh(<span class="code-quote">'script/deploy'</span>)
  }
}
</pre>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=arthurschreiber">arthurschreiber</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-33151" target="_blank">GitHub commit status design is bad</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: github-branch-source-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2016-02-25T14:14:43+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 11
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>The latest release of the GitHub Branch Source Plugin added support for github commit status notifications. I was seriously happy about this, because I thought this would allow me to remove generic </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">step([$class: <span class="code-quote">'GitHubSetCommitStatusBuilder'</span>])
</pre>
</div></div>
<p>and</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">step([$class: <span class="code-quote">'GitHubCommitNotifier'</span>, resultOnFailure: <span class="code-quote">'FAILURE'</span>])
</pre>
</div></div>
<p>steps from my workflow scripts.</p>

<p>Unfortunately, I was seriously let down by the new functionality, and I found major issues with how it interacts with the workflow plugin.</p>

<h4><a name="Nosupportformultiple%60node%60%2F%60checkout%60combinations"></a><b>No support for multiple `node` / `checkout` combinations</b></h4>

<p>  My Workflow scripts reserve different nodes and checkout the same code on
  these. This might happen in parallel, or in sequence. After the steps on the
  first node finish, the build gets marked as successful, even if later steps
  are still pending. If a later step fails, a previously successful build will
  suddenly get marked as successful. I haven't tested with parallel builds yet,
  but I believe that might cause even weirder issues.</p>

<h4><a name="Noflexibility"></a>No flexibility</h4>

<p>  My workflow jobs sometimes consist of multiple different steps,
  like Validation, Build, Test, Deployments. With the generic steps mentioned
  above, I was able to be flexible on when the "final" commit status was set.
  E.g for one job, I want the commit status to be set after the code was build
  and tested, irrelevant of the deployment step.</p>

<p>I thus propose an alternative solution:</p>

<p>Instead of making the sending of the commit status automatic, custom workflow
steps should be added to the `github` plugin. These steps should be flexible
enough to allow users to use them in even the most complex workflow scripts,
while also having sensible defaults to make them easy to use.</p>

<h4><a name="%7B%7BwithGitHubCommitStatus%7D%7D"></a><tt>withGitHubCommitStatus</tt></h4>

<p>  This step should take a block and update the commit status on GitHub. By default,
  it should use the GitHub URL, Commit ID and other required information from
  the main git source for the job.</p>

<p>  This step could take a bunch of optional parameters:</p>

<ul>
	<li><tt>sha</tt>: Commit ID to set the status for. (Default: Taken from build)</li>
	<li><tt>repo</tt>: Repository name. (Default: Taken from build)</li>
	<li><tt>owner</tt>: Owner name (user or organization). (Default: Taken from build)</li>
	<li><tt>context</tt>: Context to be used for the status. (Default: "default" or "continuous-integration/jenkins")</li>
	<li><tt>targetUrl</tt>: URL to be linked to in the GitHub UI. (Default: Build URL)</li>
</ul>

<p>  Based on whether the contained block passes or not, the status on GitHub should be set
  accordingly.</p>

<h4><a name="%7B%7BsetGitHubCommitStatus%7D%7D"></a><tt>setGitHubCommitStatus</tt></h4>

<p>  This is basically a non-block version the same as <tt>withGitHubCommitStatus</tt>.</p>

<p>  It should take the same parameters as <tt>withGitHubCommitStatus</tt>, plus the following:</p>

<ul>
	<li><tt>state</tt>: <tt>pending</tt>, <tt>success</tt>, <tt>error</tt>, or <tt>failure</tt>.</li>
	<li><tt>description</tt>: A custom description.</li>
</ul>

<p>Here's same example Workflow scripts that show how this could be used:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">#!/usr/bin/env groovy

stage name: <span class="code-quote">'build'</span>

parallel lint: {
  withGitHubCommitStatus context: <span class="code-quote">"continuous-integration/jenkins/lint"</span> {
    node {
      checkout scm

      sh(<span class="code-quote">'script/setup'</span>)
      sh(<span class="code-quote">'script/lint'</span>)
    }
  }
}, test: {
  withGitHubCommitStatus context: <span class="code-quote">"continuous-integration/jenkins/test"</span> {
    node {
      checkout scm

      sh(<span class="code-quote">'script/setup'</span>)
      sh(<span class="code-quote">'script/test'</span>)
    }
  }
}

<span class="code-comment">// Prepare a deployment
</span>stage name: <span class="code-quote">'prepare-deployment'</span>
node {
  checkout scm

  sh(<span class="code-quote">'script/setup'</span>)
  sh(<span class="code-quote">'script/prepare-deploy'</span>)
}

<span class="code-comment">// Run deployments
</span><span class="code-keyword">if</span> (currentBuild.result != <span class="code-quote">'FAILURE'</span> &amp;&amp; env.BRANCH_NAME == <span class="code-quote">'master'</span>) {
  input <span class="code-quote">'Trigger Deployment?'</span>

  stage name: <span class="code-quote">'deployment'</span>, concurrency: 1

  node {
    checkout scm

    sh(<span class="code-quote">'script/setup'</span>)
    sh(<span class="code-quote">'script/deploy'</span>)
  }
}
</pre>
</div></div></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>github-branch-source 1.2</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-33151?focusedId=249079&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-249079">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The design of multibranch projects is intentionally kept simple and focused on the most common CI tasks: one repository of any significance, all maintenance automatic. If you have more complex scenarios you can use lower-level tools such as the <a href="https://github.com/jenkinsci/github-integration-plugin" class="external-link" target="_blank" rel="nofollow noopener">GitHub Integration plugin</a> to define customized behaviors from standard Pipeline jobs. (You can also use the Job DSL plugin to create a series of jobs based on branches detected in one or more repositories.)</p>
</pre>
</details>

</ul>
</details>

<p>The design of multibranch projects is intentionally kept simple and focused on the most common CI tasks: one repository of any significance, all maintenance automatic. If you have more complex scenarios you can use lower-level tools such as the <a href="https://github.com/jenkinsci/github-integration-plugin" class="external-link" target="_blank" rel="nofollow noopener">GitHub Integration plugin</a> to define customized behaviors from standard Pipeline jobs. (You can also use the Job DSL plugin to create a series of jobs based on branches detected in one or more repositories.)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>arthurschreiber:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-33151?focusedId=249082&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-249082">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Huh, okay. I don't quite see how "automatic maintenance" and my proposal really clash.</p>

<p>Can there then at least be an option to disable the Commit status notifier? At least I'll then be able to continue using the github branch source plugin without having it send random status values to GitHub.</p>
</pre>
</details>

</ul>
</details>

<p>Huh, okay. I don't quite see how "automatic maintenance" and my proposal really clash.</p>

<p>Can there then at least be an option to disable the Commit status notifier? At least I'll then be able to continue using the github branch source plugin without having it send random status values to GitHub.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/recena.png" title="recena's avatar" /> recena:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-33151?focusedId=249098&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-249098">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=arthurschreiber" class="user-hover" rel="arthurschreiber">arthurschreiber</a> That option can be interesting. Please, file a new issue for that.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=arthurschreiber" class="user-hover" rel="arthurschreiber">arthurschreiber</a> That option can be interesting. Please, file a new issue for that.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nloewen:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-33151?focusedId=255090&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-255090">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> The plugin you linked seems to be focused exclusively on pull requests and doesn't seem to enable these kind of workflows. Additionally, I have been unable to find any plugin which allows this functionality.</p>

<p>I would find the functionality <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=arthurschreiber" class="user-hover" rel="arthurschreiber">arthurschreiber</a> described very helpful; in particular, the ability to set the context of the GitHub status. I suspect there are others who would as well. Is there any way this functionality could be reconsidered even in part?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> The plugin you linked seems to be focused exclusively on pull requests and doesn't seem to enable these kind of workflows. Additionally, I have been unable to find any plugin which allows this functionality.</p>

<p>I would find the functionality <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=arthurschreiber" class="user-hover" rel="arthurschreiber">arthurschreiber</a> described very helpful; in particular, the ability to set the context of the GitHub status. I suspect there are others who would as well. Is there any way this functionality could be reconsidered even in part?</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/github-branch-source-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-35453%22">JENKINS-35453</a>]</i>
